src/blob_file_builder.cc

```diff
@@ -39,11 +59,81 @@ void BlobFileBuilder::Add(const BlobRecord& record, BlobHandle* handle) {
   }
 }
 
+void BlobFileBuilder::EnterUnbuffered() {
+  // Using collected samples to train the compression dictionary
+  // Then replay those records in memory, encode them to blob file
+  // When above things are done, transform builder state into unbuffered
+  std::string samples = "";
+  std::vector<size_t> sample_lens;
+
+  for (const auto& rec_str : sample_records_) {
+    size_t copy_len =
+        cf_options_.blob_file_compression_options.zstd_max_train_bytes > 0
+            ? std::min(cf_options_.blob_file_compression_options
+                               .zstd_max_train_bytes -
+                           rec_str.size(),
+                       rec_str.size())
+            : rec_str.size();
+    samples.append(rec_str, 0, copy_len);
+    sample_lens.emplace_back(copy_len);
+  }
+  std::string dict;
+  dict = ZSTD_TrainDictionary(
+      samples, sample_lens,
+      cf_options_.blob_file_compression_options.max_dict_bytes);
+
+  compression_dict_.reset(
+      new CompressionDict(dict, cf_options_.blob_file_compression,
+                          cf_options_.blob_file_compression_options.level));
+  encoder_.SetCompressionDict(*compression_dict_);
+
+  builder_state_ = BuilderState::kUnbuffered;
+
+  // add history buffer
+  BlobHandle handle;
+  for (const std::string& rec_str : sample_records_) {
+    BlobRecord rec;
+    Slice rec_slice(rec_str);
+    rec.DecodeFrom(&rec_slice);
+    Add(rec, &handle);
+  }
+  sample_records_.clear();
+  sample_str_len_ = 0;
+}
+
+void BlobFileBuilder::WriteRawBlock(const Slice& block, BlockHandle* handle) {
+  handle->set_offset(file_->GetFileSize());
+  handle->set_size(block.size());
+  status_ = file_->Append(block);
```

Comment:
Let's follow rocksdb's block based table format, add 1 byte compression type and 4 bytes checksum to the end of each block. For compression type, we can use `kNoCompression` for both of the index block and the dictionary block, and leave the compression type byte as a placeholder. For the checksum let's just do crc32.

https://github.com/facebook/rocksdb/blob/master/table/block_based/block_based_table_builder.cc#L1054

Reply:
sure, I've added them after meta index block and dictionary block.